### ISSUE TITLE
Remove the CAC1 driver (another issues with card detection)

### ISSUE THREAD

**Issue by Jakuje** [OpenSC/OpenSC#1377]
### Problem Description

TL;DR: Remove CAC1 driver. It breaks detection of misbehaving cards (soon after reset?).

I run the [CI on master](https://gitlab.com/redhat-crypto/OpenSC/). We have a machine with two-slot reader with CAC and coolkey cards inserver. It worked fine until last week. The cards got swapped last week and now the coolkey card is no longer detected from my testsuite, while it works just fine if:

 * The `OPENSC_DRIVER` is set to `coolkey`
 * Configuration sets `card_drivers = coolkey, cac`
 * From `pkcs11-tool` (which is opening only one session, not more as a testsuite)

The `PKCS11SPY` trace looks like this (while before there was one more iteration of `C_Initialize()`, `C_Finalize()` before, whicih could have messed up something):
```
15: C_GetSlotInfo
2018-06-04 05:58:05.473
[in] slotID = 0x0
[out] pInfo: 
      slotDescription:        'OMNIKEY CardMan (076B:3022) 3021'
                              ' 00 00                          '
      manufacturerID:         '                                '
      hardwareVersion:         0.0
      firmwareVersion:         0.0
      flags:                   7
        CKF_TOKEN_PRESENT                
        CKF_REMOVABLE_DEVICE             
        CKF_HW_SLOT                      
Returned:  0 CKR_OK

16: C_OpenSession
2018-06-04 05:58:06.196
[in] slotID = 0x0
[in] flags = 0x6
pApplication=(nil)
Notify=(nil)
[out] *phSession = 0x16785f0
Returned:  0 CKR_OK

17: C_GetMechanismList
2018-06-04 05:58:06.196
[in] slotID = 0x0
Returned:  224 CKR_TOKEN_NOT_PRESENT
```

From the logs, there are two separate issues:

 * Misdetection of coolkey cards as CAC (again)
 * Breaking some data structures while issuing several cycles of Initialize/Finalize (?)

I can isolate the first issue to the configuration, where CAC driver is in front of coolkey driver, where I start seeing detection issues (the `pkcs11-tool` still detects the card, but in testsuite, it shows without any objects), but not yet this complete brokeness.

This is again issue of the coolkey cards responding to arbitrary APDUs with SW = `90 00`, which is not very well handled in the card drivers yet and now it bubbled down here to the CAC driver. Not even the #1367 is solving the issue since it is falling again to the legacy CAC1 driver.

### Proposed Resolution

There are certainly issues in the CAC driver. Such as out of bounds reading. I was able to get the following logs from the master:
```
[...]
0x7f6cb8f238c0 17:31:58.387 [opensc-pkcs11] pkcs15-cac.c:356:sc_pkcs15emu_cac_init: 
cert (99 bytes):
63 3A 35 35 37 3A 73 63 5F 72 65 61 64 5F 62 69 c:557:sc_read_bi
6E 61 72 79 3A 20 63 61 6C 6C 65 64 3B 20 32 35 nary: called; 25
36 20 62 79 74 65 73 20 61 74 20 69 6E 64 65 78 6 bytes at index
20 30 00 34 31 3A 73 63 5F 6C 6F 63 6B 3A 20 72  0.41:sc_lock: r
65 74 75 72 6E 69 6E 67 20 77 69 74 68 3A 20 30 eturning with: 0
20 28 53 75 63 63 65 73 73 29 0A 00 64 5F 62 69  (Success)..d_bi
6E 61 72                                        nar
0x7f6cb8f238c0 17:31:58.387 [opensc-pkcs11] pkcs15-cert.c:370:sc_pkcs15_read_certificate: called
```
I will fix these in the coming PR.

I would also like to remove the CAC1 implementation, because it is getting problematic (it is just guesswork now, how it was supposed to work -- there is no documentation about its specification) and there is no good use for it anymore (no valid CAC cards are in use, `libcacard` is also moving away from this). Would you mind if I just remove that, or should I keep it available using either environment variable or some configuration?

Anyway, dropping the CAC1 driver resolves both issues and I see the coolkey card working again.

The changes are building on top of #1367 so I am not yet sending a PR (it would be huge mess, but the changes look somehow like this, if I use ifdefs -- I don't oppose to removing the `#ifdefed` code completely:

https://github.com/Jakuje/OpenSC/commits/drop-cac1

### Steps to reproduce

^^^ Two-slot reader, in first one coolkey card, in second CAC card. Build master (today) and run testsuite:
```
cd src/tests/p11test/ && ./p11test -s 0 -p 123456
```
### Logs

Several runs reveal many logs of card detection. I can attach them if interested.

**Comment by dengert** [comment]
You say it is a two slot reader. What is that?  Do you see the problems if you use two separate readers? 
Does it work if you reverse the cards?
Maybe it is the reader problem or how OpenSC handles the two readers.

Or does one driver not clean up the `sc_card_t` structure if it fails to match, and the next driver gets a dirty `sc_card_t` structure.  

Did it only stop working last week because the cards were switched or do you suspect some code changed?

Is the problem the CAC driver is issuing SELECT AID commands to the coolkey card and they appear to work? 

You solution appears to be turn off parts of one driver to get your driver to work.
Are the developers who wrote the CAC driver still active? 
What do they say about this?
Are there any users who still have CAC1 cards? 

This sounds similar to the problem with the Muscle applet that could not recognize a SELECT AID but interpreted it as a SELECT FILE. 

Do you have an `opensc-debug.log` showing the full selection process that is failing?

**Comment by dengert** [comment]
Is the coolkey AID really: ` 0x62, 0x76, 0x01, 0xff, 0x00, 0x00, 0x00` ?
ISO 7816-4 (2005) Section 8.2.1 APPLICATION Identifier Table 90 says:
"'0' to '9' - Reserved for backward compatibility with ISO/IEC 7812-1[3] (see annex D)"

Annex D
"In ISO/IEC 7816-5:1994, it was possible to use issuer identification numbers in application identifiers. This annex indicates the format of such AIDs.

This sounds more like the Coolkey AID should have been retired years ago, and more modern cards are not treating it as an AID, but a file.

**Comment by Jakuje** [comment]
> You say it is a two slot reader. What is that? 

OMNIKEY CardMan (076B:3022) 3021 01 00

> Do you see the problems if you use two separate readers?

I do not have separate two readers at hand to try.

> Does it work if you reverse the cards?

It worked before they were swapped. But they are located in US so I can not just come and try it now.

> Or does one driver not clean up the sc_card_t structure if it fails to match, and the next driver gets a dirty sc_card_t structure.

This is something we can try to figure out from the logs.

> Did it only stop working last week because the cards were switched or do you suspect some code changed?

The [last successful run](https://gitlab.com/redhat-crypto/OpenSC/pipelines/22414686) was on May 22nd with cards in the original position (cac, coolkey). Since then, there were some changes. It is simple to verify (with the default opensc.conf and tree from May 21st):
```
# git checkout 45e4f546c2dfb3d6d232ca55d49b007a56d3cb44
# make clean && make
# cd src/tests/p11test/
# OPENSC_DEBUG=9 ./p11test -s 0 -p redhat &> /tmp/p11test_debug.log
```
This shows the same issues as with master. See the complete log [1].

> Is the problem the CAC driver is issuing SELECT AID commands to the coolkey card and they appear to work?

Yes. The detection of the CAC1 card is based only on the answers to SELECT AID APDUs, which is the issue here.

> Are the developers who wrote the CAC driver still active?
> What do they say about this?
> Are there any users who still have CAC1 cards?

Yes. It was written by Robert Relyea two years ago. I will discuss this with him.
I don't think there are any users of CAC1 out there anymore.

> This sounds similar to the problem with the Muscle applet that could not recognize a SELECT AID but interpreted it as a SELECT FILE.

Do you have some link to this issue?

> Do you have an opensc-debug.log showing the full selection process that is failing?

This is complete log from the tree state from May 21st. It looks like it is hitting both issues:
[1] [p11test_complete.log](https://github.com/OpenSC/OpenSC/files/2071322/p11test_complete.log)

This is complete log from the `drop-cac1` branch still showing the issues after detection (probably from other drivers):
[p11test_without_cac1.log](https://github.com/OpenSC/OpenSC/files/2071353/p11test_without_cac1.log)

If I add a `card_drivers = cac, coolkey` it starts to work as expected so this means also some other card driver is mangling the context somewhere.

> Is the coolkey AID really: 0x62, 0x76, 0x01, 0xff, 0x00, 0x00, 0x00 ?

Yes, according to the [coolkey Makefile](https://github.com/dogtagpki/coolkey/blob/8351de0571cb4d031bab34d6c067cbaa4630a184/applet/Makefile#L97). If that is reserved, we should talk about changing that if it would make sense, but we would have to take it to the coolkey applet developers.

**Comment by Jakuje** [comment]
From the second log, it indeed looks like the Muscle applet is matched (from a line 15294 and several other similar blocks can be seen there):
```
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] card.c:269:sc_connect_card: trying driver 'muscle'
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] card.c:401:sc_lock: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] reader-pcsc.c:613:pcsc_lock: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] card-muscle.c:818:muscle_card_reader_lock_obtained: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] card-muscle.c:826:muscle_card_reader_lock_obtained: returning with: 0 (Success)
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] card.c:441:sc_lock: returning with: 0 (Success)
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7f092b25e9c0 03:55:17.980 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:0, INS:A4, P1:4, P2:0, data(6) 0x7f092871ed10
0x7f092b25e9c0 03:55:17.981 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: reader 'OMNIKEY CardMan (076B:3022) 3021 00 00'
0x7f092b25e9c0 03:55:17.981 [opensc-pkcs11] reader-pcsc.c:285:pcsc_transmit: 
Outgoing APDU (11 bytes):
00 A4 04 00 06 A0 00 00 00 01 01 ...........
0x7f092b25e9c0 03:55:17.981 [opensc-pkcs11] reader-pcsc.c:213:pcsc_internal_transmit: called
0x7f092b25e9c0 03:55:18.008 [opensc-pkcs11] reader-pcsc.c:294:pcsc_transmit: 
Incoming APDU (2 bytes):
90 00 ..
0x7f092b25e9c0 03:55:18.008 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.008 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.008 [opensc-pkcs11] card.c:451:sc_unlock: called
0x7f092b25e9c0 03:55:18.008 [opensc-pkcs11] reader-pcsc.c:663:pcsc_unlock: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] apdu.c:554:sc_transmit_apdu: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] card.c:401:sc_lock: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] reader-pcsc.c:613:pcsc_lock: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] card-muscle.c:818:muscle_card_reader_lock_obtained: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] card-muscle.c:826:muscle_card_reader_lock_obtained: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] card.c:441:sc_lock: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] apdu.c:521:sc_transmit: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:B0, INS:3C, P1:0, P2:0, data(0) (nil)
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: reader 'OMNIKEY CardMan (076B:3022) 3021 00 00'
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] reader-pcsc.c:285:pcsc_transmit: 
Outgoing APDU (5 bytes):
B0 3C 00 00 40 .<..@
0x7f092b25e9c0 03:55:18.017 [opensc-pkcs11] reader-pcsc.c:213:pcsc_internal_transmit: called
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] reader-pcsc.c:294:pcsc_transmit: 
Incoming APDU (2 bytes):
6C 10 l.
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] apdu.c:401:sc_set_le_and_transmit: called
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] apdu.c:371:sc_single_transmit: called
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] apdu.c:378:sc_single_transmit: CLA:B0, INS:3C, P1:0, P2:0, data(0) (nil)
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] reader-pcsc.c:284:pcsc_transmit: reader 'OMNIKEY CardMan (076B:3022) 3021 00 00'
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] reader-pcsc.c:285:pcsc_transmit: 
Outgoing APDU (5 bytes):
B0 3C 00 00 10 .<...
0x7f092b25e9c0 03:55:18.093 [opensc-pkcs11] reader-pcsc.c:213:pcsc_internal_transmit: called
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] reader-pcsc.c:294:pcsc_transmit: 
Incoming APDU (18 bytes):
01 01 01 04 00 00 13 88 00 00 0B 03 01 04 00 00 ................
90 00                                           ..
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] apdu.c:390:sc_single_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] apdu.c:422:sc_set_le_and_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] apdu.c:543:sc_transmit: returning with: 0 (Success)
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] card.c:451:sc_unlock: called
0x7f092b25e9c0 03:55:18.159 [opensc-pkcs11] reader-pcsc.c:663:pcsc_unlock: called
0x7f092b25e9c0 03:55:18.166 [opensc-pkcs11] card.c:283:sc_connect_card: matched: MuscleApplet
```
The pkcs15 emulation fails to match after this, because it is based on the match above (line 20896):
```
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] pkcs15-syn.c:156:sc_pkcs15_bind_synthetic: trying coolkey
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] pkcs15-coolkey.c:725:sc_pkcs15emu_coolkey_init_ex: called
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] pkcs15-coolkey.c:66:coolkey_detect_card: called
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] pkcs15-coolkey.c:732:sc_pkcs15emu_coolkey_init_ex: returning with: -1413 (Unsupported card)
```
as well as the muscle select file later on (line 20902):
```
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] card.c:740:sc_select_file: called; type=1, path=e828bd080fa000000167455349474e::
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] card-muscle.c:377:select_item: returning with: -1201 (File not found)
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] card-muscle.c:446:muscle_select_file: returning with: -1201 (File not found)
0x7f092b25e9c0 03:55:22.665 [opensc-pkcs11] card.c:762:sc_select_file: 'SELECT' error: -1201 (File not found)
```

So it is looking like a misdetection by muscle driver (probably because the reason you mentioned above) and no other driver mangling the context (that is good).

**Comment by dengert** [comment]
The Muscle issue I was referring to starts at: https://github.com/OpenSC/OpenSC/pull/1316#issuecomment-383251983

In this case the driver a card-sc-hsm.c is willing to accept a SELECT AID response without an FCI, and Javacard do not see the command as the special SELECT AID but as a SELECT FILE which Javacard passed on to the Muscle applet that just responds 9000  without returning an FCI as expected by a SELECT AID command.  The driver did not check that a valid FCI containing the expected AID was returned. 

From your analysis above, it looks like the card-muscle.c should also be checking if a FCI with the expected AID is returned. But it may also be the case that a valid Muscle applet may not return an FCI.

ISO 7816-4  is hard to read  and uses "may" in too many places making it hard to see what is really required. 

`muscle.c` ` msc_select_applet` does `sc_format_apdu(card, &apdu, SC_APDU_CASE_3_SHORT, 0xA4, 4, 0);` and `apdu.resplen = 0;` So this might confuse a card's OS  in to not treating it as a SELECT AID but pass it to the applet that treats it as a SELECT FILE. 

Can you test with your configuration if changing  `msc_select_applet` to request a response (even if not checked) makes a difference?  i.e. does the return code remain 9000 or does it indicate the AID not found.

**Comment by Jakuje** [comment]
> Can you test with your configuration if changing msc_select_applet to request a response (even if not checked) makes a difference? i.e. does the return code remain 9000 or does it indicate the AID not found.

To my observation, it does not have any effect. The initial detection (line 481 from the second log) fails as expected with muscle applet as well as the second one (line 3562 -- second card?), but after that and after reset (?), the muscle is matched (line 7610).

If I try to drop the muscle too (by making the `muscle_match()` return 0, I am able to continue with the detection and finally match the coolkey applet.

What stuck my attention was the reference to the bug in the muscle. The coolkey is based on the Muscle so it contains very similar code (even checking the CLA against the INS values?):
https://github.com/dogtagpki/coolkey/blob/8351de0571cb4d031bab34d6c067cbaa4630a184/applet/src/com/redhat/ckey/applet/CardEdge.java#L3184

But it does not answer the question why it works in initial detection, but not soon after reset or whatever is going on there during the re-detection.

**Comment by dengert** [comment]
I agree the code in [musclecard](https://github.com/JavaCardOS/MuscleCard-Applet/blob/master/musclecard/src/com/musclecard/CardEdge/CardEdge.java)  lines 326-327 
and  [coolkey](https://github.com/dogtagpki/coolkey/blob/8351de0571cb4d031bab34d6c067cbaa4630a184/applet/src/com/redhat/ckey/applet/CardEdge.java) lines 3184-3185 looks wrong. 

Both, first test if the cardOS (Javacard) has determined the command is a SELECT AID for this applet;
Muscle lines 321 and 323, Coolkey 3167-3267:
```
if (selectingApplet())
			ISOException.throwIt(ISO7816.SW_NO_ERROR);
``` 

So to get passt the above, Javacard must have detrermined the command was not to select the applet. And the command is sent to the applet:
Muscle in lines 326 and 327 checks  for Select (A4) which includes SELECT AID as a subset:
```
if ((buffer[ISO7816.OFFSET_CLA] == 0) && (buffer[ISO7816.OFFSET_INS] == (byte) 0xA4))
			return;
```
And I contend that the return becomes a 90 00.

Coolkey line 3184 appears to do a return for any command with CLA=00.
Line 3185 is questionable, as it is testing for CLA=A4 
Lines 3183-3186
```
	switch (cla) {
	case ISO7816.CLA_ISO7816:
	case ISO7816.INS_SELECT:  // right value, but right define?
	    return;
```
So this comes down to how does Javacard determine if a command is a SELECT AID and does it
handle it internally?

Any Javacard experts reading this?



I believe that the Javacard will only catch a SELECT command if it is `00 A4 04 00` and is requesting a response length. In any other case it is just passed to the applet as a command. 

Does it only do this when no AID is selected? 

In any case it looks like both of the cards are not handling SELECT AID correctly. 

What other OpenSC supported cards are based on Javacard?  Is source available to look at?
 
If OpenSC will continue to work with Javacard applets that misbehave, we need to understand how to make sure they do not give false positive identifications of other applets.

**Comment by Jakuje** [comment]
Unfortunately, I am not a JavaCard expert, but your analysis seems logical, but it does not answer the question why the card behaves different  for the first connection than for the second.

I finally read through the latest debug logs properly (with disabled muscle driver as a broken one -- (aren't we going to remove/disable them?)[https://github.com/OpenSC/OpenSC/projects/2]) to come to the following:

 * Cryptoki is initialized using `C_Initialize()`
 * First, the slots 0 - 3 are detected: `connect_card()` matches correctly coolkey driver and is followed by successful match in `pkcs15_bind_synthetic()` and the card initialization
 * Second, the slots 4-7 are detected: `connect_card()` matches correctly PIV driver and is followed by successful match in `pkcs15_bind_synthetic()` and the card Initialization
 * The Cryptoki is reset by `C_Finalize()` and opened again
 * The slots 0 - 3 are detected: `connect_card()` matches incorrectly Smart Card HSM driver, based on the `90 00` answer from the card. Follows the `pkcs15_bind_synthetic()`, which fails during initialization of `sc-hsm` driver, which leaves the state of the card unmatched and disconnects the card (with #1382 othrewise it just leaves the structures hanging there).
 * The slots 4 - 7 are properly detected the same way as in previous step.
 * The Cryptoki is reset by `C_Finalize()` and opened again
 * Now, the **first** scenario repeats -- the coolkey card is detected -- the card was probably not reset recently so it had some time to come to "alive" state and answers correctly to APDUs
 * The Cryptoki is reset by `C_Finalize()` and opened again
 * This round again mismatches the card.

So what we can learn from here?

 * The `C_Finalize()` is reseting the card somehow, if it was detected even after we set the the `disconnect_action = leave`?
 * The misdetection of JavaCards early after reset/boot/whatever-is-that-called is still issue
 * Actually, it does not look like multi-reader/multi-card issue at all as I am getting some more understanding now -- the previous configuration was firstly detecting the PIV card so the coolkey had enough time to boot and respond correctly ... sigh ... sorry for initial confusion. I can obviously reproduce the same with single reader and single card on my laptop. I was probably not running the testsuite for some time with coolkey card.

Disabling `muscle` and `sc-hsm` card matching seems to make the testsuite pass for me.

**Comment by dengert** [comment]
"The slots 0 - 3 are detected: connect_card() matches incorrectly Smart Card HSM driver, based on the 90 00 answer from the card." Is the problem seen in https://github.com/OpenSC/OpenSC/pull/1316#issuecomment-383251983

I still think if the the drivers where to check for an FCI in the response from SELECT AID they would not get a false positive.

**Comment by Jakuje** [comment]
Yes. But since I don't have any `muscle` nor `sc-hsm` card to test, I have no way how to fix these drivers ...

**Comment by CardContact** [comment]
I still don't understand how this issue is related to the sc-hsm driver. If a card responds 9000 to a SELECT(AID) APDU, then the driver should safely assume, that the application is present on the card and was selected. That's the whole purpose of application selection in ISO 7816-4 and the standard is very specific on that.

Even on a JavaCard, the default behavior is that a SELECT(AID) with an AID not listed in the card managers registry will result in a File Not Found response. And the SELECT(AID) APDU can be clearly identified by P1=04 and can be safely distinguished from a SELECT APDU directed at an EF.

It's also common practice to validate APDUs and return appropriate SW1/SW2 for unknown combinations of CLA,INS,P1,P2. There is no logical reason to return 9000, unless the applet actually does something reasonable with the command.

It's the wrong path to change unrelated card drivers just to fix bugs in other cards.

A generally recommended way is to use unique historical bytes in the ATR, so that you can identify the card without doing application selection. This is what we do with the SmartCard-HSM - you can tell from the ATR that you have a SmartCard-HSM in front of you. However there are some cards in circulation (namely a few MicroSD cards or contactless cards in bogus reader that don't return the ATR) that can no be detected this way. For this case probing with SELECT(AID) probing is used.

**Comment by Jakuje** [comment]
> A generally recommended way is to use unique historical bytes in the ATR, so that you can identify the card without doing application selection.

Isn't the ATR assigned by the card manufacturer? With JavaCards and just custom applets loaded in there, I don't think we have any control of what ATR is used and if somebody else would load different applet to the same card, we would see mismatches.

Anyway, this was already discussed in different issues, that selecting card driver based on the ATR does not fly if there can be multiple applets on the card (Yubikeys with PIV and OpenPGP) so this does not seem as a future-proof solution. This is also a reason why this should not be the only detection mechanism.

Yes, the issue is here, that the card drivers, if they depend on the detection based on some APDUs checks, should check their answers already in the `match()` function (or even the p11card init() might work too). If the card is supposed to answer with something, and/or if there is some card-specific APDU that are supposed to return some specific buffers, such as the PIV, CAC and many other drivers do now.

What I see in `sc-hsm`, these checks come only in the PKCS15 layer `init()` function (?), where it is already too late to fall back to different card driver. Is there something like that, what could be used in the sc-hsm? I can reproduce the issue reliably so if you will have something I can verify its functionality against java card.

**Comment by CardContact** [comment]
I agree on the ATR issue and my point is, that it should be used for "product" smart cards like the SmartCard-HSM, which just happens to be a JavaCard, but where we control applets on the card and can define a product specific ATR. So we know in advance if a card with a certain ATR is a candidate for the sc-hsm driver.

What makes things complicated, is that some contactless reader don't report the ATR and that some old MicroSD cards use a default JCOP ATR. Only for those few cards we rely on the SELECT(AID) probing. What is also part of the picture is, that when the SmartCard-HSM applet is selected, then the authentication state is reset. Unfortunately OpenSC contains some weird mechanisms that trigger SELECT(AID) and thereby reset the authentication state. We already have code in the driver that prevents SELECT(AID) to preserve the authentication state, so adding another detection mechanism would need to be tested in combination with this workaround.

What we could check, is that SELECT(AID) returns a valid FCP during probing. That should be a sound indication unless some other cards (or the ISO layer) comes up with the idea to return a dummy FCP as well.

**Comment by dengert** [comment]
"Isn't the ATR assigned by the card manufacturer? "

Yes and no.

**Comment by dengert** [comment]
"What makes things complicated, is that some contactless reader don't report the ATR and that some old MicroSD cards use a default JCOP ATR. Only for those few cards we rely on the SELECT(AID) probing."

Actually, the way I read the OpenSC  code, if any card is not in your ATR list, you fall back to using  SELECT(AID) probing.  So you will be probing  and selecting the Muscle and Coolkey cards. 

You as a card manufacture and applet developer you control both. But these days there are cards and anyone can write an applet and put it on almost any card. And can put more then one applet on the same card. So OpenSC is stuck with having to do selection by AID. 

In effect an OpenSC drivers are really an applet drivers.  Look at [pkcs11-multi-application-2](https://github.com/dengert/OpenSC/tree/pkcs11-multi-application-2) that is an experimantal mod to allow two or more card drivers (applet) drivers access to the same card via PKCS\#11. p11kit and multiple user applications can already try and use multiple applets on the same card from the same process or difference processes.

**Comment by CardContact** [comment]
You can try if 

https://github.com/CardContact/OpenSC/tree/fix_1377

solves the issue.

**Comment by dengert** [comment]
Looks like it should work. But I don't have a Muscle or Coolkey card. @Jakuje could test.

**Comment by dengert** [comment]
I have been looking at source for [musclecard](https://github.com/JavaCardOS/MuscleCard-Applet/blob/master/musclecard/src/com/musclecard/CardEdge/CardEdge.java) and [coolkey](https://github.com/elmarco/coolkey/blob/master/applet/src/com/redhat/ckey/applet/CardEdge.java) but not sure if these versions are in use in current cards. 

Both appear to not return an FCI, and thus the problems we are seeing. This also means it will be very difficult to tell the cards apart as both are very old and based on the same code, and instructions suggest that one could remove parts of the code to save space, thus there could be many versions out there. 

But support a `INS_GET_STATUS = (byte) 0x3C` command with a CLA=` CardEdge_CLA = (byte) 0xB0;` that returns the same vada and have version numbers. Muscle only supports B0 3A 00 00, where as Coolkey supports B0 3A 00 00, and B0 3A 01 00 to get some memory thing. Coolkey also supports some other operations of which these two  look interesting. `INS_GET_ISSUER_INFO = (byte)0xF6;` and INS_GET_BUILDID     = (byte)0x70;`

So it might be possible to tell these two cards apart if we wanted to do so.

 https://github.com/JavaCardOS?tab=repositories look interesting as examples of Javacard applets in general to see how they handle SELECT FILE/AID commands.

**Comment by martinpaljak** [comment]
Both should be considered old-outdated-legacy-broken applets. Unlike closed cards, there are viable options for replacement and I would still consider ditching muscle support from OpenSC, for example.

**Comment by Jakuje** [comment]
@CardContact yes, the attached patch applied on master solves the issue for me if the sc-hsm is in front of the coolkey in the drivers list. Thank you.

Actually, the coolkey applet is still developed and used, but the repository where it is worked on is the following one, if I am right:

https://github.com/dogtagpki/coolkey

I don't think the muscle is used anymore and dropping its support or at least disabling it would make sense (and was already [talked about](https://github.com/OpenSC/OpenSC/projects/2)). Would you like from me to have a look into that or did somebody already put something together?

I also confirm, that the Coolkey does not return any FCI on SELECT:
```
$ opensc-tool -s 00A4040007627601FF000000
Using reader with a card: OMNIKEY AG CardMan 3121 00 00
Sending: 00 A4 04 00 07 62 76 01 FF 00 00 00 
Received (SW1=0x90, SW2=0x00)
```
I can also confirm that other APDUs work fine with my card:
```
$ opensc-tool -s B03C000000  # INS_GET_STATUS
Using reader with a card: OMNIKEY AG CardMan 3121 00 00
Sending: B0 3C 00 00 00 
Received (SW1=0x90, SW2=0x00):
01 01 01 04 00 00 13 88 00 00 0B BC 01 04 00 00 ................
$ opensc-tool -s B070000004  # INS_GET_BUILDID
Using reader with a card: OMNIKEY AG CardMan 3121 00 00
Sending: B0 70 00 00 04 
Received (SW1=0x90, SW2=0x00):
54 DE 79 0F T.y.
$ opensc-tool -s B0F60000e0 # INS_GET_ISSUER_INFO
Using reader with a card: OMNIKEY AG CardMan 3121 00 00
Sending: B0 F6 00 00 E0 
Received (SW1=0x90, SW2=0x00):
68 74 74 70 3A 2F 2F 64 68 63 70 2D 31 36 2D 32 http://dhcp-16-2
[...]
```
Hopefully, they will work soon-after-boot too.

**Comment by Jakuje** [comment]
Thanks. I am not sure if we should not leave this open yet, since only one issue got resolved by the last commit, while at least two more were identified in the comments that are still to be handled (CAC, muscle, maybe coolkey). Or should I open separate issues for them?

**Comment by dengert** [comment]
Opening a separate issue would be a good idea. There maybe other drivers that do not check SELECT AID returned data.  Each driver should be able to distinguish its card/applet and not depend on the order of the drivers. We should not have to reorder the list of drivers.

This would be especially important if we implement something like  [pkcs11-multi-application-2](https://github.com/dengert/OpenSC/tree/pkcs11-multi-application-2)
 to allow for two applets on the same card to be usable from PKCS\#11.
